### PR TITLE
PR for Katie

### PR DESCRIPTION
**Pull request by eshipunova**
I’m speaking with myself, number one, because I have a very good brain and I’ve said a lot of things... I know what I’m doing and I listen to a lot of people, I talk to a lot of people and at the appropriate time I’ll tell you who the people are. But I speak to a lot of people. I always told people, you know I'm a very smart guy. I got good marks. 
(citation needed)

### CODE REVIEW

File: my_main.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+        finsum1, finsumm2 = split_and_print(a_number, another_number_second_part)
+        if len(str(finsum1 + finsumm2)) > 1:
+            summm = str(finsum1 + finsumm2)
+            print("{} + {} = {}".format(finsum1, finsumm2, finsum1 + finsumm2))
+            print("{} + {} = {}{}{}".format(int(summm[0]), int(summm[1]), '\033[1m', int(summm[0]) + int(summm[1]), '\033[0m'))
+        else:
+            print("{} + {} = {}{}{}".format(finsum1, finsumm2, '\033[1m', finsum1 + finsumm2, '\033[0m'))
+
+
+do_it_all(4371730)
```
Comment (cgosmeyer) [inline my_main.py:67]: Required: Encapsulate in a __main__:

```
if __name__ == "__main__:
    do_it_all(4371730)
```

File: my_main.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        # 43|71730: 4 + 3 = 7, 7 + 1 + 7 + 3 + 0 = 18 => 7 = 7, 1 + 8 = 9 => 7 + 9 = 16 => 1 + 6 = 7
+        # 437|1730: 4 + 3 + 7 = 14, 1 + 7 + 3 + 0 = 11 => 1 + 4 = 5, 1 + 1 = 2 => 5 + 2 = 7
+        # 4371|730: 4 + 3 + 7 + 1 = 15, 7 + 3 + 0 = 10 => 1 + 5 = 6, 1 + 0 = 1 => 6 + 1 = 7
+        # It's cool that it's always 7. The proof is actually in the fact that we work with a decimal system but this is a neat little demonstration that it does work
+
+        # It smells. Suspiciously so
+        # Oh no, the kettle!!
+        raise Exception
+
+def print_summm(n):
```
Comment (cgosmeyer) [inline my_main.py:30]: Suggestion: Typo for print_**sum**?  If not, might name it the function something more descriptive.

File: my_main.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
+# HMMMM what should i do
+# I think book-reading should be an easy one. Or a math computation of some kind.
+# Inputs: number
+# Outputs: all possible ways of adding digits together, printed out
+
+def print_all(n):
```
Comment (cgosmeyer) [inline my_main.py:6]: Required: Remove comment lines.

File: my_main.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        # 4 = 4, 2 + 1 = 3 =>
+        # 4 + 3 = 7
+        # 43|71730: 4 + 3 = 7, 7 + 1 + 7 + 3 + 0 = 18 => 7 = 7, 1 + 8 = 9 => 7 + 9 = 16 => 1 + 6 = 7
+        # 437|1730: 4 + 3 + 7 = 14, 1 + 7 + 3 + 0 = 11 => 1 + 4 = 5, 1 + 1 = 2 => 5 + 2 = 7
+        # 4371|730: 4 + 3 + 7 + 1 = 15, 7 + 3 + 0 = 10 => 1 + 5 = 6, 1 + 0 = 1 => 6 + 1 = 7
+        # It's cool that it's always 7. The proof is actually in the fact that we work with a decimal system but this is a neat little demonstration that it does work
+
+        # It smells. Suspiciously so
+        # Oh no, the kettle!!
+        raise Exception
```
Comment (cgosmeyer) [inline my_main.py:28]: Required: Remove Exception.

File: my_main.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    return t
+
+def get_TWO_numbers_from_ONE_number(ALL_number,split_poInt):
+    s = str(ALL_number)
+    FIRST_number = int(s[:split_poInt+1])
+    SECOND_number = int(s[split_poInt+1:])
+    return FIRST_number, SECOND_number
+
+def split_and_print(numberone,numbertwo):
+    print("{} = {}, {} = {}".format(" + ".join(map(str, str(numberone))), print_summm(numberone),
```
Comment (cgosmeyer) [inline my_main.py:42]: Required: In the join, should rename the `str` variable because it is already the name of a built-in function.

File: README.md
Comment (cgosmeyer) [inline README.md:1]: Required: The Readme should clarify what problem the code helps to solve, how to install the code, and how to run the code.

File: my_main.py
Comment (cgosmeyer) [inline my_main.py:1]: Fun toy script!

**Does it run?**
Yes! No errors.

**Structure:**

- Need to add doc strings to all the functions. See [here](https://kairosaerospace.atlassian.net/wiki/spaces/HOWTO/pages/2794520660/Python+and+GitHub+Best+Practices+for+Analysts#Best-Practices-in-Python-Documentation). 
- Need to implement a consistent naming convention. There is a mix of upper and lower cases. I would suggest going all lowercase and add more underscores where a camelcase was being used.
- Consider reorganizing so that all the functions are in a separate file (a module) and get imported into this script (i.e., it contains the `__main__`).  In other words, if this is meant to be re-used and shared with a lot of people, it needs to be reorganized into a package.  Go [here ](https://kairosaerospace.atlassian.net/wiki/spaces/HOWTO/pages/2794520660/Python+and+GitHub+Best+Practices+for+Analysts#Package-File-Structure) for more information.
- Remove all the commented out test lines. If you want to retain them for future testing, consider encapsulating them into [logging statements](https://www.loggly.com/ultimate-guide/python-logging-basics/) or in an `if test==True` block.

File: my_main.py [hunk 6]
```diff
@@ -0,0 +1,67 @@
+# HMMMM what should i do
```
Comment (cgosmeyer) [inline my_main.py:1]: Required: Remove superfluous comments.

Suggestion: If there is documentation at the top of the file, it is a Python best practice to contain it in a triple quote rather than with hashtags. For example:

```
"""
This is a script for outputting all possible additions of inputted digits.
"""
```

File: __main__.py
Comment (cgosmeyer) [inline __main__.py:1]: Suggestion: hmmm `__main__.py` is new to me but I do not think it is being used appropriately here?  I think that it should contain imports of your functions and organize them into the main script. So it essentially should take the place of the `do_it_all(4371730)` call in `my_main.py`.  See my comment in the review summary about reorganizing into a package structure. Either the `__main__.py` should be removed or , in lieu of a main script, the main call should be moved here (with the appropriate imports).

File: my_main.py (hunk 1 above)
Comment (cgosmeyer) [inline my_main.py:67]: Required: This number should not be hard-coded, but could be used in a unit-test or in an example for the Readme.

File: my_main.py [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+    numberone = print_summm(numberone)
+    numbertwo = print_summm(numbertwo)
+    if len(str(numberone)) > 1 or len(str(numbertwo)) > 1:
+        #print(len(str(numberone)), len(str(numbertwo)))
+        split_and_print(numberone, numbertwo)
+    return print_summm(numberone), print_summm(numbertwo)
+
+def do_it_all(n):
+    n = str(n)
+    for i in range(len(n)-1):   # for 4371730 it would be 0,1,2,3,4,5
```
Comment (cgosmeyer) [inline my_main.py:54]: Required: Consider renaming unclear or inconsistent variables `finsumm2` (typo?), `summm` (difficult to read), and `and a_number` and `another_number_second_part` (names could be more self-explanatory).

Suggestion: It is not immediately clear what the code is doing.  More descriptive variable and function names would help. In addition to such self-documentation tactics, it might be helpful to add comment lines.

File: my_main.py [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+        # It smells. Suspiciously so
+        # Oh no, the kettle!!
+        raise Exception
+
+def print_summm(n):
+    s = str(n)
+    t = sum([int(d) for d in s])
+    return t
+
+def get_TWO_numbers_from_ONE_number(ALL_number,split_poInt):
```
Comment (cgosmeyer) [inline my_main.py:35]: Required: Lower case all the upper case to retain a consistent style.

**Review by cgosmeyer — COMMENTED** [review]
Fun toy script!

**Functionality:**

- I like the lack of library imports. No fuss to set up.
- Runs without errors - but only tested the hard-coded number.
- Suggestion: Consider[ adding command line arguments](https://docs.python.org/3/library/argparse.html) so that users can input numbers from terminal rather than editing the main in the script.
- Suggestion: Consider creating [unit tests ](https://docs.python.org/3/library/unittest.html)for each of the functions. Need to show that it won't break for other numbers or edge cases. 
- Suggestion: Add [exceptions](https://docs.python.org/3/tutorial/errors.html) to catch errors when users input blanks or other edge cases known to cause issues.

**Structure:**

- Suggestion: Consider reorganizing so that all the functions are in a separate file (a module) and get imported into this script (i.e., it contains the `__main__`).  In other words, if this is meant to be re-used and shared with other people, it needs to be reorganized into a package.  Go [here ](https://kairosaerospace.atlassian.net/wiki/spaces/HOWTO/pages/2794520660/Python+and+GitHub+Best+Practices+for+Analysts#Package-File-Structure) for more information.
- Suggestion: With a package structure in mind, and after splitting `my_main.py` into two files (a module and a script), the script should be renamed to something more descriptive than `my_main`.
- Required: Remove all the commented out test lines. If you want to retain them for future testing, consider encapsulating them into [logging statements](https://www.loggly.com/ultimate-guide/python-logging-basics/) or in an `if test==True` block.

**Style:**

- Required: Need to add doc strings to all the functions. See [here](https://kairosaerospace.atlassian.net/wiki/spaces/HOWTO/pages/2794520660/Python+and+GitHub+Best+Practices+for+Analysts#Best-Practices-in-Python-Documentation). 
- Required: Need to implement a consistent naming convention. For example, there is a mix of upper and lower cases.

File: my_main.py (hunk 2 above)
Comment (eshipunova) [reply my_main.py:30]: I love the distinction between suggestions and requirements!

File: my_main.py (hunk 5 above)
Comment (eshipunova) [reply my_main.py:42]: This is actually not a variable, see `Using map() with a string as an iterator` in https://www.guru99.com/python-map-function.html. In this case, redundantly, I am applying `str` as a function. The reason I am doing that is so I can have a list of individual digits to join together

File: README.md
Comment (eshipunova) [reply README.md:1]: Really good point!

**Comment by eshipunova** [comment]
LOVE the suggestions, thoughtful and on point. Really appreciate the distinction between suggested and required, as well as calling out the good parts of a script. You made really thoughtful comments about file structure, and suggested future improvements. Your summary can be used as a how-to example!

What could be added to this review: what happens with different inputs? what breaks?
